common/test/acceptance/tests/helpers.py

```diff
@@ -354,6 +354,80 @@ def verify_events_of_type(self, username, event_type, expected_events, expected_
                     "Refer '{0}' does not have correct suffix, '{1}'.".format(actual_referer, expected_referers[index])
                 )
 
+    def create_event_promise(self, check_event_func, description, **kwargs):
+        """
+        Return a promise that is fulfilled when an event results in `check_event_func(event)` returning True.
+
+        Note that `check_event_func` will be called on all events emitted after this moment in time.
+        """
+        return EventPromise(
+            self.event_collection,
+            self.start_time,
+            check_event_func,
+            description,
+            **kwargs
+        )
+
+    def create_event_of_type_promise(self, event_type):
+        """
+        Return a promise that is fulfilled when the first event of the requested type is emitted.
+
+        Note that this does not de-duplicate events, so code like this will not do what you expect::
+
+            promise_a = self.create_event_of_type_promise('foobar')
+            promise_b = self.create_event_of_type_promise('foobar')
+            event_a = promise_a.fulfill()
+            event_b = promise_b.fulfill()
+
+            assert event_a == event_b  # this is true! both of these promises return the first matching event!
+
+        If you want to distinguish between the two "foobar" events it is recommended that you use `create_event_promise`
+        and specify a more rigorous function for checking if you have found the correct event.
+
+        This function is designed for cases when only one event of a particular type will be emitted on a page.
+
+        """
+        def check_event_type(event):
+            return event.get('event_type', '') == event_type
+
+        return self.create_event_promise(
+            check_event_type,
+            'Waiting for a {0} event to be emitted'.format(event_type),
+        )
+
+
+class EventPromise(Promise):
```

Comment:
I like the idea of using arbitrary patterns to capture a set of events. I'm not yet convinced that a subclass of Promise is the way to go, but I'm not fundamentally opposed.

Here are some thoughts:
 - generally the condition used for the promise should also be used to return the matching events for testing. Given this, I would think a helper method that takes a function would be a more useful abstraction, as it can use the same matching function both in a promise and in returning the results.
 - Bok Choy tests usually hide promises inside the PageObject classes, so that the tests are written without concern for asynchronous browser behavior. This isn't exclusively true, but I only see a few counterexamples, and IMO they would all be better factored this way too.  I like acceptance tests to be very concise and human readable, and to match up clearly with the BDD that we're verifying.
 - we found when writing our event tests that it was very useful to have an expected array of events and to use assertEquals against the actual events. This gives an excellent failure message with the diff of expected vs actual JSON, which made it very easy to see what was actually captured and how it differed. See this example:

https://github.com/edx/edx-platform/blob/master/common/test/acceptance/tests/lms/test_account_settings.py#L269
 - It would be great if you have a flexible event matching mechanism to update the other helper methods to use it too. Again though, the key to me is that the tests are readable, so I wouldn't want the tests themselves to have too much complex matching logic. I'd prefer to have that complexity live in the page object classes.

I hope this is helpful.